### PR TITLE
persist state for wicked, chrony, and systemd

### LINKED ISSUE #248: DHCP leases should be persisted on local storage
Presently `wicked` saves lease info to `/run`, where it will be lost on reboot.

This avoids needing to trust data from persistent storage, but can introduce an availability issue.

If we're confident that wicked is sufficiently hardened to malicious input from this source after code review, we can move the data back to `/var`.

### LINKED ISSUE #249: chrony drift files should be persisted on local storage
Presently `chrony` saves its drift files to `/run`, where they will be lost on reboot.

This avoids needing to trust data from persistent storage, but can introduce an availability issue.

If we're confident that `chrony` is sufficiently hardened to malicious input from this source after code review, we can move the data back to `/var`.

### LINKED ISSUE #250: systemd journal and machine ID should be persisted on local storage
Right now systemd is defaulting to "volatile" journal behavior, because `/var/log` doesn't exist when the journal starts.

In order for the journal to effectively persist, we also need a way to persist the machine ID across restarts. This is awkward without an initramfs; we'd need to give `preinit` some way of accessing the ID before anything else is up.

I'd like to handle this without introducing an initramfs, but if we end up with one for other reasons, it would be trivial to fix.

- [ ] persist machine ID across restarts
- [ ] persist journal across restarts

### PR DESCRIPTION
**Pull request by bcressey**
**Issue number:**
#248 , #249 , #250 

**Description of changes:**
Confine chrony, wicked, and dbus-broker via SELinux policy; move DHCP leases and clock measurements back to persistent storage. Also persist the journal and the random seed.

**Testing done:**
No AVC denials logged in `dmesg` or in the journal.

Processes are labeled as expected.
```
system_u:system_r:bus_t:s0         3023 ?        Ss     0:00 /usr/bin/dbus-broker-launch --scope system
system_u:system_r:bus_t:s0         3067 ?        S      0:00 dbus-broker --log 4 --controller 9 --machine-id 990978842db9412194d093dc75af968b --max-
system_u:system_r:network_t:s0     3124 ?        Ss     0:00 /x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/wicked/bin/wickedd-auto4 --systemd
system_u:system_r:network_t:s0     3136 ?        Ss     0:00 /x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/wicked/bin/wickedd-dhcp4 --systemd
system_u:system_r:network_t:s0     3155 ?        Ss     0:00 /x86_64-bottlerocket-linux-gnu/sys-root/usr/libexec/wicked/bin/wickedd-dhcp6 --systemd
system_u:system_r:network_t:s0     3218 ?        Ss     0:00 /x86_64-bottlerocket-linux-gnu/sys-root/usr/sbin/wickedd --systemd --foreground
system_u:system_r:network_t:s0     3280 ?        Ss     0:00 /x86_64-bottlerocket-linux-gnu/sys-root/usr/sbin/wickedd-nanny --systemd --foreground
system_u:system_r:clock_t:s0       3555 ?        Ss     0:00 /usr/sbin/chronyd -d -F -1
system_u:system_r:clock_t:s0       3610 ?        S      0:00 /usr/sbin/chronyd -d -F -1
```

Files are labeled as expected.
```
# ls -latrZ /var/lib/wicked
-rwx------. 1 root root system_u:object_r:lease_t:s0   38 Jul  7 17:39 iaid.xml
-rwx------. 1 root root system_u:object_r:lease_t:s0   55 Jul  7 17:39 duid.xml
-rw-------. 1 root root system_u:object_r:lease_t:s0  996 Jul  7 20:03 lease-eth0-dhcp-ipv4.xml

# ls -latrZ /var/lib/chrony/
-rw-r--r--. 1 chrony chrony system_u:object_r:measure_t:s0 1121 Jul  7 20:09 169.254.169.123.dat
-rwx------. 1 chrony chrony system_u:object_r:measure_t:s0   42 Jul  7 20:09 drift

# ls -latrZ /var/lib/systemd/
-rwxr-xr-x.  1 root root system_u:object_r:state_t:s0  512 Jul  8 14:51 random-seed
drwxr-xr-x.  2 root root system_u:object_r:state_t:s0 4096 Jul  8 14:51 coredump
drwxr-xr-x.  2 root root system_u:object_r:state_t:s0 4096 Jul  8 14:51 catalog

# ls -latrZ /var/log/journal/ec274cd5add60d0d045b205a4a4317a8/
-rw-r-----. 1 root systemd-journal system_u:object_r:state_t:s0 8388608 Jul  8 14:58 system.journal
```

The machine ID is consistent across reboots and stop/start for both m4.xlarge and m5.xlarge.
```
/* m4.xlarge */
# cat /sys/hypervisor/uuid
ec274cd5-add6-0d0d-045b-205a4a4317a8

# cat /etc/machine-id
ec274cd5add60d0d045b205a4a4317a8

/* m5.xlarge */
# cat /etc/machine-id
ec2e505a672c97f59ba650eb81e9ed6d

# cat /sys/class/dmi/id/product_uuid
ec2e505a-672c-97f5-9ba6-50eb81e9ed6d
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Review by tjkirch — APPROVED** [review]

File: packages/systemd/9001-move-stateful-paths-to-ephemeral-storage.patch [hunk 1]
```diff
@@ -1,45 +0,0 @@
...
--                *ret = "/var/lib";
-+                *ret = "/run/lib";
-                 return 0;
- 
-         case SD_PATH_SYSTEM_STATE_LOGS:
--                *ret = "/var/log";
-+                *ret = "/run/log";
-                 return 0;
- 
-         case SD_PATH_SYSTEM_STATE_CACHE:
```
Comment (jahkeup) [inline packages/systemd/9001-move-stateful-paths-to-ephemeral-storage.patch:32]: Is there an expected set of use cases that suggest we keep `/var/cache` around? I'm wondering if reconfigurations and upgrades may have implicit conflicts with data under this path that we can avoid by *not* persisting `/var/cache`.

File: packages/systemd/var-run-tmpfiles.conf [hunk 2]
```diff
@@ -0,0 +1,4 @@
+d /run/cache 0755 root root -
```
Comment (jahkeup) [inline packages/systemd/var-run-tmpfiles.conf:1]: Is `/run/cache` symlinked at `/var/cache` by another process or config? Or should this set of tmpfiles rules include an entry for `/var/cache`?

File: packages/systemd/9001-move-stateful-paths-to-ephemeral-storage.patch (hunk 1 above)
Comment (tjkirch) [reply packages/systemd/9001-move-stateful-paths-to-ephemeral-storage.patch:32]: migrator uses `/var/cache/bottlerocket-metadata` after reboot.

File: packages/systemd/9004-machine-id-setup-generate-stable-ID-under-Xen.patch [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                                 log_info("Initializing machine ID from KVM UUID.");
+                                 return 0;
+                         }
++
++                } else if (detect_vm() == VIRTUALIZATION_XEN) {
++
++                        /* Otherwise, if we're running under Xen, check for
++                         * a UUID from the hypervisor. */
++
++                        if (id128_read("/sys/hypervisor/uuid", ID128_UUID, ret) >= 0) {
```
Comment (tjkirch) [inline packages/systemd/9004-machine-id-setup-generate-stable-ID-under-Xen.patch:40]: This is unique to each instance, not to the hypervisor?

File: packages/systemd/9001-move-stateful-paths-to-ephemeral-storage.patch (hunk 1 above)
Comment (bcressey) [reply packages/systemd/9001-move-stateful-paths-to-ephemeral-storage.patch:32]: We don't have any uses of `/var/cache` today outside of updog and migrator.

We have a `/var/cache/private` that's created for use with the `CacheDirectory` directive in unit files, but we don't use that feature.

File: packages/systemd/var-run-tmpfiles.conf (hunk 2 above)
Comment (bcressey) [reply packages/systemd/var-run-tmpfiles.conf:1]: `/run/cache` is used by our updates API, and also for the ldconfig cache.

`/var/cache` is separate. It gets created as a side effect of the tmpfiles snippet for `/var/cache/private`, but would also be created by updog if it wasn't present.

**Review by tjkirch — COMMENTED** [review]
I was left confused after Jake's comment - we're not losing `/var/cache` on reboot, right?  We rely on it for migrations, at least, and surely other things.

The testing focuses on SELinux, and does show the machine ID, but you confirmed that all these things are persisting over reboot?  Old journal entries, same machine ID, chrony timings, wicked lease...

**Comment by bcressey** [comment]
> I was left confused after Jake's comment - we're not losing `/var/cache` on reboot, right? We rely on it for migrations, at least, and surely other things.

Correct, `/var/cache` is not going away.

> The testing focuses on SELinux, and does show the machine ID, but you confirmed that all these things are persisting over reboot? Old journal entries, same machine ID, chrony timings, wicked lease...

Yes, I've confirmed that the files are persisted across reboot.

**Review by tjkirch — APPROVED** [review]

File: packages/systemd/9004-machine-id-setup-generate-stable-ID-under-Xen.patch (hunk 3 above)
Comment (tjkirch) [reply packages/systemd/9004-machine-id-setup-generate-stable-ID-under-Xen.patch:40]: Ben confirmed that it is unique to each instance; the hypervisor sees all zeroes.

**Comment by bcressey** [comment]
I've also tested a downgrade from a system with the stateful files into an older version. wicked and chrony were both stateless before and didn't expect files in `/var/lib`, so those were fine.

My concern was with journald, which previously defaulted to `Storage=auto`, which uses a persistent journal if `/var/log/journal` exists, and otherwise uses `/run/log/journal`. Since the persistent journal is created by the upgrade, it will still be there on the downgrade, and labeled with an SELinux label that didn't exist in the older policy.

It ended up working fine, though, since the policy allows for writes to files with no label or an unknown label.

**Review by jamieand — APPROVED** [review]
LGTM

**Pull request merged**